pkg/gardenlet/controller/seed/seed_lease_control.go

```diff
@@ -0,0 +1,139 @@
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	seed, err := c.seedLister.Get(name)
+	if apierrors.IsNotFound(err) {
+		logger.Logger.Infof("[SEED LEASE] Stopping lease operations for Seed %s since it has been deleted", key)
+		c.seedLeaseQueue.Done(key)
+		return nil
+	}
+	if err != nil {
+		logger.Logger.Infof("[SEED LEASE] unable to retrieve Seed object %s from store: %v", key, err)
+		return err
+	}
+
+	if err := c.checkSeedConnection(seed); err != nil {
+		return fmt.Errorf("[SEED LEASE] cannot establish connection with Seed %s: %v", key, err)
+	}
+
+	condition := helper.GetCondition(seed.Status.Conditions, gardencorev1beta1.SeedGardenletReady)
+	if condition == nil {
+		condition, err = c.initGardenletReadyCondition(seed)
+		if err != nil {
+			return err
+		}
+	}
+
+	seedOwnerReference := buildSeedOwnerReference(seed)
+	if err := c.seedLeaseControl.Sync(seed.GetName(), *seedOwnerReference); err != nil {
+		return err
+	}
+
+	if condition.Status == gardencorev1beta1.ConditionUnknown {
+		err := c.updateGardenletReadyCondition(seed)
+		if err != nil {
+			return err
+		}
+	}
+
+	durationAfter := time.Duration(leaseResyncSeconds * time.Second)
+	c.seedLeaseQueue.AddAfter(key, durationAfter)
+	return nil
+}
+
+func buildSeedOwnerReference(seed *gardencorev1beta1.Seed) *metav1.OwnerReference {
+	return &metav1.OwnerReference{
+		APIVersion: gardencorev1beta1.SchemeGroupVersion.WithKind("Seed").Version,
+		Kind:       gardencorev1beta1.SchemeGroupVersion.WithKind("Seed").Kind,
+		Name:       seed.GetName(),
+		UID:        seed.GetUID(),
+	}
+}
+
+func (c *Controller) initGardenletReadyCondition(seed *gardencorev1beta1.Seed) (*gardencorev1beta1.Condition, error) {
+	conditionGardenletReady := helper.InitCondition(gardencorev1beta1.SeedGardenletReady)
+	conditionGardenletReady = helper.UpdatedCondition(conditionGardenletReady, gardencorev1beta1.ConditionTrue, "GardenletReady", "Gardenlet is posting ready status.")
+	err := c.updateSeedStatus(seed, conditionGardenletReady)
+	return &conditionGardenletReady, err
+}
+
+func (c *Controller) updateGardenletReadyCondition(seed *gardencorev1beta1.Seed) error {
+	var conditionGardenletReady = helper.GetOrInitCondition(seed.Status.Conditions, gardencorev1beta1.SeedGardenletReady)
+	conditionGardenletReady = helper.UpdatedCondition(conditionGardenletReady, gardencorev1beta1.ConditionTrue, "GardenletReady", "Gardenlet is posting ready status.")
+	return c.updateSeedStatus(seed, conditionGardenletReady)
+}
+
+func (c *Controller) updateSeedStatus(seed *gardencorev1beta1.Seed, conditionGardenletReady gardencorev1beta1.Condition) error {
+	_, err := kutil.TryUpdateSeedConditions(c.k8sGardenClient.GardenCore(), retry.DefaultBackoff, seed.ObjectMeta,
+		func(seed *gardencorev1beta1.Seed) (*gardencorev1beta1.Seed, error) {
+			seed.Status.Conditions = helper.MergeConditions(seed.Status.Conditions, conditionGardenletReady)
+			return seed, nil
+		},
+	)
+	return err
+}
```

Comment:
The methods here are making update requests to the API server:

1. For objects which don't need update
1. With retry. Retry logic should be handled by workqueue.